### PR TITLE
CodeQL docs: update titles, add intros, and a few content updates

### PR DESCRIPTION
**Pull request by jf205**
The PR makes a start at addressing the tasks in https://github.com/github/product-documentation/issues/2163.

**tldr for any members of the C analysis team who stumble upon this**: We are trying to align the CodeQL documentation with the style and structure used in the GitHub help docs. There will be a series of PRs with updates. We'll try not to do anything that could change the technical meaning of the text and may ping you with questions as we are going along.

### Summary of changes
 
- Update titles
- Update links to account for new titles
- Add intros
- Rework the `CodeQL for C and C++` homepage to include only the autogenerated TOC.
- Some small changes to content:
  - Attempt to make the flow from the new intro to the main article a bit smoother
  - Fix a few inline links

### TODO

Possibly next PR:
- Rework further reading sections, and save in a reusable snippet as suggested by @felicitymay in #2823. I'll wait until that PR is merged to avoid conflicts.
- More thorough review for localization issues.

Later PRs:
- Think long and hard about the most appropriate content models to apply.
- Review content, adding referential, conceptual, and procedural sections in certain topics.

For more info, see notes on https://github.com/github/product-documentation/issues/2163.

### CODE REVIEW

File: docs/language/learn-ql/cpp/introduce-libraries-cpp.rst [hunk 1]
```diff
@@ -1,20 +1,25 @@
...
 .. code-block:: ql
 
    import cpp
 
 The rest of this topic summarizes the available CodeQL classes and corresponding C/C++ constructs.
 
-NOTE: You can find related classes and features using the query console's auto-complete feature.  You can also press *F3* to jump to the definition of any element; library files are opened in new tabs in the console.
+.. pull-quote:: Note
+
+   You can find related classes and features using the query console's auto-complete feature.  You can also press *F3* to jump to the definition of any element. Library files are opened in new tabs in the console.
```
Comment (felicitymay) [inline docs/language/learn-ql/cpp/introduce-libraries-cpp.rst]: Not required for this PR, but this note should probably have been deleted or clarified when we moved the documentation out of LGTM. We should probably check other languages for similar notes.

File: docs/language/learn-ql/cpp/private-field-initialization.rst [hunk 2]
```diff
@@ -1,13 +1,15 @@
-Example: Checking that constructors initialize all private fields
-=================================================================
+Refining a query to account for edge cases
+==========================================
+
+You can improve the results generated by a CodeQL query by adding conditions to remove false positives caused by common edge cases.
```
Comment (felicitymay) [inline docs/language/learn-ql/cpp/private-field-initialization.rst]:
```suggestion
You can improve the results generated by a CodeQL query by adding conditions to remove false positive results caused by common edge cases.
```

File: docs/language/learn-ql/cpp/private-field-initialization.rst [hunk 3]
```diff
@@ -124,7 +126,7 @@ This case can be excluded by creating a recursive predicate. The recursive predi
 Refinement 4—simplifying the query
 ----------------------------------
 
-Finally we can simplify the query by using the `transitive closure operator <https://help.semmle.com/QL/ql-handbook/recursion.html#transitive-closures>`__. In this final version of the query, ``c.calls*(fun)`` resolves to the set of all functions that are ``c`` itself, are called by ``c``, are called by a function that is called by ``c``, and so on. This eliminates the need to make a new predicate all together.
+Finally we can simplify the query by using the transitive closure operator. In this final version of the query, ``c.calls*(fun)`` resolves to the set of all functions that are ``c`` itself, are called by ``c``, are called by a function that is called by ``c``, and so on. This eliminates the need to make a new predicate all together. For more information, see `transitive closures <https://help.semmle.com/QL/ql-handbook/recursion.html#transitive-closures>`__ in the QL language handbook.
```
Comment (shati-patel) [inline docs/language/learn-ql/cpp/private-field-initialization.rst]:
```suggestion
Finally we can simplify the query by using the transitive closure operator. In this final version of the query, ``c.calls*(fun)`` resolves to the set of all functions that are ``c`` itself, are called by ``c``, are called by a function that is called by ``c``, and so on. This eliminates the need to make a new predicate all together. For more information, see `Transitive closures <https://help.semmle.com/QL/ql-handbook/recursion.html#transitive-closures>`__ in the QL language handbook.
```

File: docs/language/learn-ql/cpp/value-numbering-hash-cons.rst [hunk 4]
```diff
@@ -1,15 +1,14 @@
...
+In C and C++ databases, each node in the abstract syntax tree is represented by a separate object. This allows both analysis and results display to refer to specific appearances of a piece of syntax. However, it is frequently useful to determine whether two expressions are equivalent, either syntactically or semantically.
 
+The hash consing library (defined in ``semmle.code.cpp.valuenumbering.HashCons``) provides a mechanism for identifying expressions that have the same syntactic structure. The global value numbering library (defined in ``semmle.code.cpp.valuenumbering.GlobalValueNumbering``) provides a mechanism for identifying expressions that compute the same value at runtime. Both libraries partition the expressions in each function into equivalence classes represented by objects. Each ``HashCons`` object represents a set of expressions with identical parse trees, while ``GVN`` objects represent sets of expressions that will always compute the same value. For more information, see `hash consing <https://en.wikipedia.org/wiki/Hash_consing>`__ and `value numbering <https://en.wikipedia.org/wiki/Value_numbering>`__ on Wikipedia.
```
Comment (shati-patel) [inline docs/language/learn-ql/cpp/value-numbering-hash-cons.rst]:
```suggestion
The hash consing library (defined in ``semmle.code.cpp.valuenumbering.HashCons``) provides a mechanism for identifying expressions that have the same syntactic structure. The global value numbering library (defined in ``semmle.code.cpp.valuenumbering.GlobalValueNumbering``) provides a mechanism for identifying expressions that compute the same value at runtime. Both libraries partition the expressions in each function into equivalence classes represented by objects. Each ``HashCons`` object represents a set of expressions with identical parse trees, while ``GVN`` objects represent sets of expressions that will always compute the same value. For more information, see `Hash consing <https://en.wikipedia.org/wiki/Hash_consing>`__ and `Value numbering <https://en.wikipedia.org/wiki/Value_numbering>`__ on Wikipedia.
```

File: docs/language/learn-ql/cpp/zero-space-terminator.rst [hunk 5]
```diff
@@ -98,7 +100,7 @@ When you have defined the basic query then you can refine the query to include f
 Improving the query using the 'SSA' library
 -------------------------------------------
 
-The ``SSA`` library represents variables in `static single assignment <http://en.wikipedia.org/wiki/Static_single_assignment_form>`__ (SSA) form. In this form, each variable is assigned exactly once and every variable is defined before it is used. The use of SSA variables simplifies queries considerably as much of the local data flow analysis has been done for us.
+The ``SSA`` library represents variables in static single assignment (SSA) form. In this form, each variable is assigned exactly once and every variable is defined before it is used. The use of SSA variables simplifies queries considerably as much of the local data flow analysis has been done for us. For more information, see `static single assignment <http://en.wikipedia.org/wiki/Static_single_assignment_form>`__ on Wikipedia.
```
Comment (shati-patel) [inline docs/language/learn-ql/cpp/zero-space-terminator.rst]:
```suggestion
The ``SSA`` library represents variables in static single assignment (SSA) form. In this form, each variable is assigned exactly once and every variable is defined before it is used. The use of SSA variables simplifies queries considerably as much of the local data flow analysis has been done for us. For more information, see `Static single assignment <http://en.wikipedia.org/wiki/Static_single_assignment_form>`__ on Wikipedia.
```

**Comment by jf205** [comment]
@shati-patel, @felicitymay@: thanks for the comments and suggestions. Is there anything else you think I should address now? (I intend to revisit this when all of the other initial 'CodeQL for X'  PRs have been completed, as mentioned on #2866 and #2823.)

**Comment by felicitymay** [comment]
Thanks for the updates 👍

**Comment by jf205** [comment]
I removed the `WIP` label now. Any further thoughts before i merge?

**Comment by felicitymay** [comment]
I think this is fine to merge as-is.
